### PR TITLE
2-stage docker build

### PR DESCRIPTION
**Pull request by whatnick**
Add dockerfile updates while keeping previous OWS stuff.

### CODE REVIEW

File: dockerfile [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
-USER root
+COPY requirements-docker.txt /
+RUN /usr/local/bin/env-build-tool new /requirements-docker.txt /env
 
-# Install C version of PyYaml
-RUN apt-get update && apt-get install -y --no-install-recommends libyaml-dev && \
-    rm -rf /var/lib/apt/lists/* && \
-    pip3 --no-cache-dir install --verbose --force-reinstall -I pyyaml
+FROM opendatacube/geobase:runner
+COPY --from=env_builder /env /env
```
Comment (Kirill888) [inline dockerfile]: I recommend copying environment AFTER all the `apt` stuff, purely due to docker caching layers nonsense, usually python env changes more often then your extra system requirements

File: dockerfile [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
 
-# Ensure compatible versions of boto are installed
-RUN pip3 --no-cache-dir install -U 'aiobotocore[awscli,boto3]' google-cloud-storage
+# Environment can be whatever is supported by setup.py
+# so, either deployment, test
+ARG ENVIRONMENT=deployment
+RUN echo "Environment is: $ENVIRONMENT"
+
+# Do the apt install process, including more recent Postgres/PostGIS
+RUN apt-get update && apt-get install -y wget gnupg \
```
Comment (Kirill888) [inline dockerfile]: either put all your apt install in docker (more common approach) or put all of them in your other file

File: requirements-apt.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+libyaml-dev
```
Comment (Kirill888) [inline requirements-apt.txt]: do you really need build-essential and yaml dev? Not if you install everything in build stage.

File: dockerfile [hunk 4]
```diff
@@ -1,16 +1,42 @@
...
+    && sed 's/#.*//' /tmp/requirements-apt.txt | xargs apt-get install -y \
+    && rm -rf /var/lib/apt/lists/*
+
+ENV PATH="/env/bin:${PATH}"
+
+# Set up a nice workdir, and only copy the things we care about in
+RUN mkdir -p /code
+ADD . /code
+
+# Install Python requirements
```
Comment (Kirill888) [inline dockerfile]: why not do this in builder stage?

File: requirements-apt.txt (hunk 3 above)
Comment (whatnick) [reply requirements-apt.txt]: Not sure what this means. libyaml-dev is there to ensure presence of the C-yaml linbrary. If this is already in the builder we can skip it.

File: dockerfile (hunk 2 above)
Comment (whatnick) [reply dockerfile]: The wget and gpupg enable install of the pg-11 client library to better support pg-11 servers.

**Comment by whatnick** [comment]
via @alexgleith :

```
Pedantic: dockerfile should be Dockerfile
For the requirements, copying all three requirements files in at once would be better, and then running the pip installs in a single RUN command will be much better.
It doesn't need the Environment can be whaterver... section.
The COPY of the update_ranges etc should probably go into /usr/local/bin/ so that they can be called without the path... I think that'll be nice :slightly_smiling_face: . Ideally they'll not be dodgy wrappers too.
```

**Pull request merged**